lib/oli/authoring/editing/container_editor.ex

```diff
@@ -264,6 +266,102 @@ defmodule Oli.Authoring.Editing.ContainerEditor do
     end
   end
 
+  @doc """
+    Duplicates a page within its project.
+    A true deep-copy of all the content is made. Resources and Revisions are created instead of simply referenced.
+  """
+  def duplicate_page(
+        %Revision{} = container,
+        page_id,
+        %Author{} = author,
+        %Project{} = project
+      ) do
+    original_page =
+      Resources.get_revision!(page_id)
+      |> Map.from_struct()
+
+    new_page_attrs =
+      original_page
+      |> Map.delete(:slug)
+      |> Map.delete(:inserted_at)
+      |> Map.delete(:updated_at)
+      |> Map.delete(:updated_at)
+      |> Map.delete(:resource_id)
+      |> Map.delete(:resource)
+      |> Map.put(:title, "Copy of #{original_page.title}")
+      |> Map.put(:content, nil)
+
+    Repo.transaction(fn ->
+      with {:ok, created_revision} <- add_new(container, new_page_attrs, author, project),
+           {:ok, duplicated_content} <-
+             duplicate_content_activities(
+               original_page.content,
+               project.slug,
+               author
+             ),
+           {:ok, updated_revision} <-
+             Resources.update_revision(created_revision, %{content: duplicated_content}) do
+        updated_revision
+      else
+        {:error, e} -> Repo.rollback(e)
+      end
+    end)
+  end
+
+  defp duplicate_content_activities(content, project_slug, author) do
+    case deep_copy_activities(content["model"], project_slug, author) do
+      {:ok, duplicated_activities} ->
+        {:ok, %{content | "model" => Enum.reverse(duplicated_activities)}}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  defp deep_copy_activities([], _project_slug, _author), do: {:ok, []}
+
+  defp deep_copy_activities(model, project_slug, author) do
+    Enum.reduce_while(model, {:ok, []}, fn activity, {:ok, acc} ->
```

Comment:
Got it, although just to clarify, this implementation does not duplicate adaptive pages. And this is one of the reasons why we are implementing an adaptive page duplication separately, alongside different content formats, making it easier to debug/extend, etc.

You probably have more context on the amount data currently handled in production, do you think it's still necessary to implement some sort of streamlined/batch creation for normal graded and non-graded pages?

Initially thought the ROI might not be super high, however learning how the objectives need to be handled I probably have to avoid using `ActivityEditor.create` altogether. Which means I can probably go for that either way, which makes sense.